### ISSUE TITLE
Remove dependency on JSR305 annotations

### ISSUE THREAD

**Issue by KengoTODA** [spotbugs/spotbugs#421]
SpotBugs implementation depends on JSR305 (e.g. `TypeQualifierAnnotation.java`) so we need to change implementation to remove dependency on outdated JSR305.

refs #180

**Comment by rschmitt** [comment]
Will support for JSR-305 annotations be removed altogether, or will they still be optionally supported?

**Comment by KengoTODA** [comment]
SpotBugs core also depends on JSR305, so we need to remove it too.

**Comment by earcam** [comment]
Hi,

While I understand that JSR305 != FindBugs/SpotBugs, I don't understand the statement 

> JSR305 is already Dormant status, so SpotBugs does not release jsr305 jar file. Please continue using findbugs’ one.

(from http://spotbugs.readthedocs.io/en/latest/migration.html)

Given that the JSR305 annotations are used to annotation FindBugs/SpotBugs, doesn't it make more sense to take ownership and port these to a new namespace?  Or do I not see the complete picture (is this about the meta/validator side of things or legal?)

It's a little depressing to see com.google.code.findbugs:jsr305:3.0.2 now relegated to a transitive dependency that I cannot... must not use 😉

I'll really miss `@javax.annotation.WillClose`, `@javax.annotation.ParametersAreNonnullByDefault` and friends - it would be great to have equivalents somewhere.

Thanks

**Comment by KengoTODA** [comment]
> Given that the JSR305 annotations are used to annotation FindBugs/SpotBugs, doesn't it make more sense to take ownership and port these to a new namespace?

Please send your PR! I think nobody blocks you to make it real :)

**Comment by earcam** [comment]
@KengoTODA, I've done the preliminary work in a couple of repos:

- https://github.com/earcam/spotbugs/tree/jsr-305
- https://github.com/earcam/jsr-305

Built with `gradlew build` and all tests are passing.   The jsr-305 project is refactored as follows:

| Old                                      | New |
|------------------------------------|-------|
|javax.annotation                   | com.github.spotbugs.jsr305.annotation |
|javax.annotation.concurrent | com.github.spotbugs.jsr305.annotation.concurrent |
|javax.annotation.meta          | com.github.spotbugs.jsr305.annotation.meta |

I'm happy to continue working on the jsr-305 to get it into a release-able state (including OSGi manifest and JPMS module-info with JDK1.6 target), but (as indicated by using spotbug's namespace) I'd rather this was owned by SpotBugs long term.

Note:  I think it makes sense to still support the FindBugs JSR-305 annotations - the forked branch of SpotBugs largely does that, there are a couple of places in the code that would need updating, these are marked with a `//FIXME` comment.

Does this sound OK?  (if so, PRs on the way)

**Comment by ThrawnCA** [comment]
The trouble with taking ownership and changing the namespace is that [various tools already recognise `javax.annotations` despite it being dormant](https://sling.apache.org/documentation/development/jsr-305.html). Changing the package name loses those tools.

Is there any way to continue development within the `javax.annotations` package?

**Comment by KengoTODA** [comment]
Note: JSR305 ri (reference implementation) is licensed under BSD-3-Clause.

* https://github.com/amaembo/jsr-305/blob/master/ri/LICENSE

**Comment by KengoTODA** [comment]
> Is there any way to continue development within the javax.annotations package?

I know one solution: [TypeQualifierNickname](https://static.javadoc.io/com.google.code.findbugs/jsr305/3.0.2/javax/annotation/meta/TypeQualifierNickname.html). But it means that our implementation still needs to depends on `javax.annotation` implementation, at least on compile time.

**Comment by ThrawnCA** [comment]
So...should we be coordinating with @amaembo then? He's a former FindBugs developer IIUC. Are there particular changes that JSR-305 needs, which we might be able to get done?

**Comment by KengoTODA** [comment]
> should we be coordinating with @amaembo then? 

No. Type qualifier nickname is a feature to handle annotation as nickname/facade for another annotation. For instance, [Nullable](https://static.javadoc.io/com.google.code.findbugs/jsr305/3.0.2/javax/annotation/Nullable.html) is nickname for [Nonnull](https://static.javadoc.io/com.google.code.findbugs/jsr305/3.0.2/javax/annotation/Nonnull.html).

**Comment by KengoTODA** [comment]
So I thought that we can annotate our new annotations with `TypeQualifierNickname`, then tools can treat them as a nickname for JSR305 annotations.

**Comment by ThrawnCA** [comment]
> Nullable is nickname for Nonnull

Um...that's quite contrary to its Javadoc btw. `@Nullable` is equivalent to having no annotation at all.

> we can annotate our new annotations with TypeQualifierNickname, then tools can treat them as a nickname for JSR305 annotations

OK...what is the expected benefit of having the `com.github.spotbugs` versions? If the annotations need development, then we'll want to backport that to the JSR305 ones, right?

**Comment by earcam** [comment]
The issue is that FindBugs JSR305 causes a split package with module [java.xml.ws.annotation](https://docs.oracle.com/javase/9/docs/api/java.xml.ws.annotation-summary.html) which creates a [world of pain](http://openjdk.java.net/jeps/261#EE-modules) for Java >= 9

Releasing javax.* without Oracle approval is a license violation AFAIK.

There is possibly another option - ownership of `javax.annotation` namespace now comes under the EE4J project maintained by Eclipse.

I've submitted a [PR for common-annotations-api](https://github.com/eclipse-ee4j/common-annotations-api/pull/5) which would resolve this issue nicely.

**Comment by alexei-osipov** [comment]
> I've submitted a [PR for common-annotations-api](https://github.com/eclipse-ee4j/common-annotations-api/pull/5) which would resolve this issue nicely.

It's unlikely that this PR will be ever merged because it does not resolve the "split package" problem with build-in Java module "java.xml.ws.annotation" (example of conflicting class: `javax.annotation.PostConstruct`).

**Comment by xenoterracide** [comment]
+1 to any kind of fix for this, SuppressFBWarnings doesn't compile due to the missing `java.annotations` Just use a different implementation, there's 3 or 4 now

**Comment by alexei-osipov** [comment]
@xenoterracide It does not look like the issue is going to be fixed anytime soon.

So by now the best option is just to replace SpotBugs annotations with any other set of annotations that does not depend on JSR305 as long it covers desired set of functionality.

You can see list of alternatives compatible with SoptBugs tool here: [NullnessAnnotation.java](https://github.com/spotbugs/spotbugs/blob/cfc3597e04ea175876ff3545ee1a6fb558ed5753/spotbugs/src/main/java/edu/umd/cs/findbugs/ba/NullnessAnnotation.java#L39-L48)

**Comment by xenoterracide** [comment]
except I'm not doing null checking with spotbugs, I'm trying to ignore a different false positive, but using these annotations simply results (somehow) on it being needed at compile time.

**Comment by hazendaz** [comment]
In some cases, I'm running into need to exclude this.  I think its time we look at removing this in 5.0.0 and there are a number of other already in works efforts to bring in newer annotation support at same time.  I suspect its also noted here or elsewhere as i didn't re-read this but Oracle is getting pretty picky on their copyright again so its better we comply rather than note regardless.  We already have spotbugs annotations so we just need to work this out.  If someone wants to take up trying to get a pull request together, it would be much appreciated.

**Comment by hazendaz** [comment]
> worth saying that spring, gradle, and more are starting to use jspecify but I'm not certain that that's relevant to this issue at all.

Yes, last I checked they will be doing so in spring 7.  That is enough to make it mainstream and the right path forward in general.  As a result, other projects I work on are also now picking it up simply because spring has agreed to do so.
